### ISSUE TITLE
cyrus-imapd crash on UID STORE with empty list (debian version 3.4.2-1~bpo11+1)

### ISSUE THREAD

**Issue by mildred** [cyrusimap/cyrus-imapd#3787]
When syncinc with imapsync, I found out that Cyrus was crashing on the destination side. See https://github.com/imapsync/imapsync/issues/293#issuecomment-973847546

I'm not too sure about what happens on the IMAP connection, but if I trust imapsync logs, the last IMAP command should be:

```
4101 UID STORE 683 FLAGS.SILENT ()
```

The system logs for cyrus in the meantime is:

```
Nov 19 08:10:40 os-gravelines-1.webmsg.me cyrus/imap[81561]: login: localhost [127.0.0.1] <<<REDACTED>>> plaintext User logged in SESSIONID=<cyrus-1637309440-81561-1-1138792220851950945>
Nov 19 08:10:40 os-gravelines-1.webmsg.me cyrus/imap[81561]: TLS is available.: func=<tls_enabled>
Nov 19 08:10:40 os-gravelines-1.webmsg.me cyrus/imap[81561]: fetching user_deny.db entry for '<<<REDACTED>>>'
Nov 19 08:10:40 os-gravelines-1.webmsg.me cyrus/imap[81561]: command: 4099 Select
Nov 19 08:10:40 os-gravelines-1.webmsg.me cyrus/imap[81561]: open: user <<<REDACTED>>> opened INBOX
Nov 19 08:10:40 os-gravelines-1.webmsg.me cyrus/imap[81561]: fetching user_deny.db entry for '<<<REDACTED>>>'
Nov 19 08:10:40 os-gravelines-1.webmsg.me cyrus/imap[81561]: command: 4100 Uid
Nov 19 08:10:40 os-gravelines-1.webmsg.me cyrus/imap[81561]: fetching user_deny.db entry for '<<<REDACTED>>>'
Nov 19 08:10:40 os-gravelines-1.webmsg.me cyrus/imap[81561]: command: 4101 Uid
Nov 19 08:10:40 os-gravelines-1.webmsg.me cyrus/imap[81561]: auditlog: touched sessionid=<cyrus-1637309440-81561-1-1138792220851950945> mailbox=<mildred.fr!user.mildred> uniqueid=<j840zwemf4i4kcgi8gzjvuwe> uid=<683> guid=<73475bb0c07a245d70aa2b497ed85069d360fe6b> cid=<e2b48251cb132f4f> modseq=<479823> oldflags=<> sysflags=<>
Nov 19 08:10:40 os-gravelines-1.webmsg.me cyrus/imap[81561]: auditlog: modseq sessionid=<cyrus-1637309440-81561-1-1138792220851950945> mailbox=<mildred.fr!user.mildred> uniqueid=<j840zwemf4i4kcgi8gzjvuwe> highestmodseq=<479823> deletedmodseq=<0>
Nov 19 08:10:40 os-gravelines-1.webmsg.me cyrus/imap[81561]: sync_rightnow_connect((null))
Nov 19 08:10:40 os-gravelines-1.webmsg.me cyrus/master[36954]: process type:SERVICE name:imap path:/usr/lib/cyrus/bin/imapd age:0.105s pid:81561 signaled to death by signal 11 (Segmentation fault)
Nov 19 08:10:40 os-gravelines-1.webmsg.me cyrus/master[36954]: service imap/ipv4 pid 81561 in BUSY state: terminated abnormally
```

I suppose that might be because the UID STORE command is passed with an empty flag list and that is not expected and that could cause empty list NULL dereferencing. Still, cyrus should not crash.

**Comment by mildred** [comment]
Reproduced it locally with imtest

```
UID FETCH 683 FLAGS
* 683 FETCH (FLAGS (\Seen))
UID OK Completed (0.000 sec)
UID STORE 683 FLAGS.SILENT ()
Connection closed.
```

**Comment by mildred** [comment]
The problem comes with storing flags to UID 683, whatever the list. The flag list is changed but right after cyrus-imapd is crashing

**Comment by mildred** [comment]
backtrace with symbols:
```
Program received signal SIGSEGV, Segmentation fault.
strlcpy (dst=dst@entry=0x55d24a2a43c0 "", src=src@entry=0x0, len=len@entry=64) at lib/xstrlcpy.c:61
61      lib/xstrlcpy.c: No such file or directory.
(gdb) bt
#0  strlcpy (dst=dst@entry=0x55d24a2a43c0 "", src=src@entry=0x0, len=len@entry=64) at lib/xstrlcpy.c:61
#1  0x00007fe5ef226e34 in backend_connect (ret_backend=ret_backend@entry=0x0, server=0x0, prot=prot@entry=0x55d248dbf400 <imap_csync_protocol>, 
    userid=userid@entry=0x55d248dae1ff "", cb=cb@entry=0x55d24a24e840, auth_status=auth_status@entry=0x7ffdb518f310, logfd=-1) at imap/backend.c:1010
#2  0x000055d248dabe3b in sync_connect (sync_cs=sync_cs@entry=0x55d248dc1fe0 <rightnow_sync_cs>) at imap/sync_support.c:7463
#3  0x000055d248dac1ee in sync_checkpoint (clientin=0x55d24a253ca0) at imap/sync_support.c:7582
#4  0x000055d248d808c3 in cmd_store (tag=0x55d24a252ed0 "UID", sequence=0x55d24a252f50 "683", usinguid=usinguid@entry=0) at imap/imapd.c:6006
#5  0x000055d248d92844 in cmdloop () at imap/imapd.c:1953
#6  0x000055d248d9a944 in service_main (argc=<optimized out>, argv=<optimized out>, envp=envp@entry=0x7ffdb51936b8) at imap/imapd.c:999
#7  0x000055d248d75e7a in main (argc=<optimized out>, argv=0x7ffdb5192550, envp=0x7ffdb51936b8) at master/service.c:647
```

**Comment by mildred** [comment]
That's here:

https://github.com/cyrusimap/cyrus-imapd/blob/4e1859e928d72b6dd746eb2a27f6643883b4b02c/imap/backend.c#L991-L1010

`backend_connect` is called with NULL `server`

Called from

https://github.com/cyrusimap/cyrus-imapd/blob/4e1859e928d72b6dd746eb2a27f6643883b4b02c/imap/sync_support.c#L7463-L7465

Where `sync_cs->servername` is NULL

**Comment by mildred** [comment]
sync_cs contains many NULL values:

```
(gdb) frame 2
#2  0x000055d248dabe3b in sync_connect (sync_cs=sync_cs@entry=0x55d248dc1fe0 <rightnow_sync_cs>) at imap/sync_support.c:7463
(gdb) print *sync_cs
$2 = {backend = 0x0, clientin = 0x0, servername = 0x0, channel = 0x55d24a1f35e0 "bourderie strasbourg", cachedb = 0x0, tagbuf = {s = 0x0, len = 0, alloc = 0, 
    flags = 0}, flags = 2}
```

This is a master Cyrus server that replicated to three two other slave servers (bourderie and stransbourg). What's strange ere is that the channel contain both channel names. Perhaps this is a misconfiguration

**Comment by mildred** [comment]
`sync_cs->servername` is not defined in this function, let's look in the parent:

https://github.com/cyrusimap/cyrus-imapd/blob/4e1859e928d72b6dd746eb2a27f6643883b4b02c/imap/sync_support.c#L7576-L7582

Because the channel name contains a space character, it will never find out the correct sync_server for both:

```
(gdb) frame 3
#3  0x000055d248dac1ee in sync_checkpoint (clientin=0x55d24a253ca0) at imap/sync_support.c:7582
7582    in imap/sync_support.c
(gdb) print rightnow_sync_cs.channel
$3 = 0x55d24a1f35e0 "bourderie strasbourg"
```

imapd.conf sync configuration:

```
sync_log: 1                                                                                                                                                             
sync_log_channels: bourderie strasbourg
sync_rightnow_channel: bourderie strasbourg
bourderie_sync_authname:admin
bourderie_sync_host:<REDACTED>
bourderie_sync_password:<REDACTED>
bourderie_sync_realm:<REDACTED>
strasbourg_sync_authname:admin
strasbourg_sync_host:<REDACTED>
strasbourg_sync_password:<REDACTED>
strasbourg_sync_realm:<REDACTED>
```

The offending configuration seems to be:
```
sync_rightnow_channel: bourderie strasbourg
```

I suppose only a single channel name is accepted

**Comment by mildred** [comment]
TL;DR, that's caused by a misconfiguration (I added a new synchronization target and regenerated the config file without mich looking at the sync_rightnow_channel setting that can only contain a single channel name)

**Comment by elliefm** [comment]
Ahh interesting.  Yeah it doesn't make any sense for `sync_rightnow_channel` to have multiple channels, but it would be nice to not crash.

Maybe we should validate it at startup and refuse to start if it's bad, or log a warning and ignore everything except the first word, or something.
